session/pingpong/provider_registrar.go

```diff
@@ -0,0 +1,212 @@
+	DelayBetweenRetries time.Duration
+	AccountantAddress   common.Address
+	RegistryAddress     common.Address
+}
+
+// NewProviderRegistrar creates a new instance of provider registrar
+func NewProviderRegistrar(transactor txer, bcHelper bc, prc ProviderRegistrarConfig) *ProviderRegistrar {
+	return &ProviderRegistrar{
+		stopChan:             make(chan struct{}),
+		bc:                   bcHelper,
+		queue:                make(chan queuedEvent),
+		registeredIdentities: make(map[string]struct{}),
+		cfg:                  prc,
+		txer:                 transactor,
+	}
+}
+
+// Subscribe subscribes the provider registrar to service state change events
+func (pr *ProviderRegistrar) Subscribe(eb eventbus.EventBus) error {
+	err := eb.SubscribeAsync(event.Topic, pr.handleNodeStartupEvents)
+	if err != nil {
+		return errors.Wrap(err, "could not subscribe to node events")
+	}
+	return eb.SubscribeAsync(service.StatusTopic, pr.consumeServiceEvent)
+}
+
+func (pr *ProviderRegistrar) handleNodeStartupEvents(e event.Payload) {
+	if e.Status == event.StatusStarted {
+		err := pr.start()
+		if err != nil {
+			log.Error().Err(err).Stack().Msgf("fatal error for provider identity registrar. Identity will not be registered. Please restart your node.")
+		}
+		return
+	}
+	if e.Status == event.StatusStopped {
+		pr.stop()
+		return
+	}
+}
+
+func (pr *ProviderRegistrar) consumeServiceEvent(event service.EventPayload) {
+	pr.queue <- queuedEvent{
+		event:   event,
+		retries: 0,
+	}
+}
+
+func (pr *ProviderRegistrar) needsHandling(qe queuedEvent) bool {
+	if qe.event.Status != string(service.Running) {
+		log.Debug().Msgf("received %q service event, ignoring", qe.event.Status)
+		return false
+	}
+
+	if _, ok := pr.registeredIdentities[qe.event.ProviderID]; ok {
+		log.Info().Msgf("provider %q already marked as registered, skipping", qe.event.ProviderID)
+		return false
+	}
+
+	return true
+}
+
+func (pr *ProviderRegistrar) handleEventWithRetries(qe queuedEvent) error {
+	err := pr.handleEvent(qe)
+	if err == nil {
+		return nil
+	}
+	if qe.retries < pr.cfg.MaxRetries {
+		log.Error().Err(err).Stack().Msgf("could not complete registration for provider %q. Will retry. Retry %v of %v", qe.event.ProviderID, qe.retries, pr.cfg.MaxRetries)
+		qe.retries++
+		go pr.delayedRequeue(qe)
+		return nil
+	}
+
+	return errors.Wrap(err, "max attempts reached for provider registration")
+}
+
+func (pr *ProviderRegistrar) delayedRequeue(qe queuedEvent) {
+	select {
+	case <-pr.stopChan:
+		return
+	case <-time.After(pr.cfg.DelayBetweenRetries):
+		pr.queue <- qe
+	}
+}
+
+func (pr *ProviderRegistrar) handleEvent(qe queuedEvent) error {
+	registered, err := pr.bc.IsRegisteredAsProvider(pr.cfg.AccountantAddress, pr.cfg.RegistryAddress, common.HexToAddress(qe.event.ProviderID))
+	if err != nil {
+		return errors.Wrap(err, "could not check registration status on BC")
+	}
+
+	if registered {
+		log.Info().Msgf("provider %q already registered on bc, skipping", qe.event.ProviderID)
```

Comment:
Use sentence case for log messages. @vkuznecovas

Reply:
k